cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java

```diff
@@ -225,74 +293,107 @@ protected void run() {
-          // messages, however, that shouldn't be persisted to avoid potential loss of messages
-          // in case of failure
-          updateTopicToRequest(topicConfig, monitorMessages);
-          continue;
-        }
-      }
+    return latestPublishTime;
+  }
 
-      String topic = requestKeyToLocalTopic.get(topicConfig);
-      latestPublishTime = Math.max(publish(topicConfig, topic, monitorMessages), latestPublishTime);
+  private void updateProgramFinishTime(Map<String, List<MonitorMessage>> monitorResponses) {
+    List<MonitorMessage> programStatusMessages = monitorResponses.get(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC);
+
+    if (programStatusMessages.isEmpty()) {
+      return;
+    }
+
+    if (programFinishTime < 0) {
+      programFinishTime = findProgramFinishTime(programStatusMessages);
+    }
+
+    if (programFinishTime >= 0) {
+      List<MonitorMessage> messages = monitorResponses.get(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC);
+      // Buffer the program state messages and don't publish them until the end
+      // Otherwise, once we publish, the deprovisioner will kick in and delete the cluster
+      // which could result in losing the last set of messages for some topics.
+      lastProgramStateMessages.addAll(messages);
+      // We still update the in memory store for the next fetch offset to avoid fetching duplicate
+      // messages, however, that shouldn't be persisted to avoid potential loss of messages
+      // in case of failure
+      updateTopicToRequest(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC, messages);
     }
-    return latestPublishTime;
   }
 
   /**
-   * Publish the given set of messages to the local TMS.
+   * Transactionally publish the given set of messages to the local TMS.
    *
    * @return the latest message publish time or {@code -1L} if there is no message to process
    */
-  private long publish(String topicConfig, String topic, List<MonitorMessage> messages) throws Exception {
-    // TODO publish messages transactionally along with offset table updates
+  private long publish(String topicConfig, String topic, List<MonitorMessage> messages,
+                       AppMetadataStore store) throws Exception {
+    // publish messages to tms
+    MessagePublisher messagePublisher = messagingContext.getMessagePublisher();
     messagePublisher.publish(NamespaceId.SYSTEM.getNamespace(), topic,
                              messages.stream().map(s -> s.getMessage().getBytes(StandardCharsets.UTF_8)).iterator());
 
-    MonitorMessage lastMessage = updateTopicToRequest(topicConfig, messages);
+    // persist the last published message as offset in meta store
+    MonitorMessage lastMessage = messages.get(messages.size() - 1);
+    store.persistSubscriberState(topicConfig, programRunId.getRun(), lastMessage.getMessageId());
+
+    // Messages have been successfully published, update in memory offset state
+    updateTopicToRequest(topicConfig, messages);
```

Comment:
If the transaction failed to commit, the in memory state has been updated. This will cause skipping of messages.